### PR TITLE
RFD 158: Description of how we should address a targeted account auth DoS

### PR DESCRIPTION
**Pull request by jentfoo**
This RFD describes how we should address the security issue captured in https://github.com/gravitational/security-findings/issues/75

The removal of this feature is already available as a PR under: https://github.com/gravitational/teleport/pull/35325

### CODE REVIEW

**Review by klizhentas — COMMENTED** [review]
@jentfoo after reading this, not clear to me 100% what are you proposing. Are you proposing not to lock out the account after trying recovery token 3 times and failing?

**Comment by jentfoo** [comment]
> Are you proposing not to lock out the account after trying recovery token 3 times and failing?

@klizhentas That is correct.  It looks like the account reset mechanisms already takes in inputs that can not be brute forced.  So I am recommending we remove this lock out in order to allow accounts that are under a targeted attack to be able to be accessed.

**Review by zmb3 — COMMENTED** [review]
There are a lot of sections in the RFD template that are not included here.

While this is a relatively targeted change and you can probably get away without all of them, I think there are some that are worth discussing.

- required approvers
- any gRPC or API changes? Any RPCs that will be deprecated by the change?
- audit/observability: should we introduce any new metrics to track account recovery attempts?

File: rfd/0158-account-recovery-protections.md
Comment (jentfoo) [reply rfd/0158-account-recovery-protections.md:16]: The password reset mechanism unlocks the account from failed password auth failures.  Which is a common pattern since theoretically the user likely entered an incorrect password repeatedly before starting the password reset flow.

**Comment by jentfoo** [comment]
> any gRPC or API changes? Any RPCs that will be deprecated by the change?

Please see the linked PR https://github.com/gravitational/teleport/pull/35325.  There is some discussion on how much we should remove from gRPC.  I don't feel strongly one way or another, so looking for your guidance 

> audit/observability: should we introduce any new metrics to track account recovery attempts?

These audit events already exist.  I feel like it's more obvious when reviewing the linked PR above.

File: rfd/0158-account-recovery-protections.md
Comment (jentfoo) [reply rfd/0158-account-recovery-protections.md]: I added a sentence here.  I was concerned about sounding repetitive considering the small focus of this RFD.  Thank you!

**Comment by jentfoo** [comment]
@zmb3 @klizhentas friendly ping on this RFD and the change https://github.com/gravitational/teleport/pull/35325

The external researcher is asking for updates on this design.  I would like to be able to respond if this is something we plan to address.

**Comment by zmb3** [comment]
@jentfoo I'm having trouble approving this because there isn't a lot of detail here. While I understand that you've already made the change, "please see the linked PR" isn't really the spirit of an RFD.

I'd expect that we use the RFD to discuss the scope of the change, and then after we've agreed upon what is in scope and what is not we can review the implementation.

**Comment by jentfoo** [comment]
@zmb3 I have added the requested sections, let me know if there is anything else I can answer.  Thank you!

**Review by zmb3 — APPROVED** [review]

File: rfd/0158-account-recovery-protections.md
Comment (strideynet) [inline rfd/0158-account-recovery-protections.md:34]: What length are reset tokens at the moment ? Do we need to increase their length if we're removing the 3 attempt limit.

**Review by strideynet — APPROVED** [review]

File: rfd/0158-account-recovery-protections.md
Comment (jentfoo) [reply rfd/0158-account-recovery-protections.md:34]: The length varies due to us using a combination of words.  Each recovery code consists of 8 words.  When considering how brute forcible this is, it's tempting to look at the character count (which on average is close to our max input length of 72 characters), but instead we need to recognize the word list is well understood.  There are 7775 possible words. which means the probability per guess is `(1/7775)^8` (result is `7.5E-32`).

This is an extremely low probability, well outside of the possibility of brute forcing.  Using a similar length fully random string would technically provide more entropy and brute force resistance, but considering the scale I don't feel that it's a necessary change.

**Comment by jentfoo** [comment]
I plan to merge this RFD next week, let me know if anyone has any additional feedback.  Thank you!

**Review by codingllama — APPROVED** [review]

**Comment by codingllama** [comment]
> I plan to merge this RFD next week, let me know if anyone has any additional feedback. Thank you!

Please make sure you have all the necessary approvals, including product.

I would suggest seeking security approval as well (apart from yourself), so you get an extra pair of eyes there too. Otherwise, happy merging ;)

File: rfd/0158-account-recovery-protections.md
Comment (jentfoo) [reply rfd/0158-account-recovery-protections.md:38]: I did not immediately notice more, but we can sort that out in the linked PR after this RFD

**Comment by jentfoo** [comment]
> Please make sure you have all the necessary approvals, including product.

My understanding is that after 2 weeks of asking for review, it is assumed to be able to be merged.  This was mentioned in the Engineering All Hands some months ago.  I still would like any additional feedback, but I have poked people several times now.

**Pull request merged**
